Add Aquacell integration

Jordi1990 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This PR adds the Aquacell integration, which can be used to monitor recent models of the Aquacell water softening device. Useful to monitor the salt levels of the device and track salt consumption over time. The API is reverse engineered from the mobile app.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: home-assistant/home-assistant.io#32664

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

Jordi1990 committed:
Initial commit

Jordi1990 committed:
Support changed API
Change sensor entity descriptions

Jordi1990 committed:
Fix sensor not handling coordinator update

Jordi1990 committed:
Implement re-authentication flow and handle token expiry

Jordi1990 committed:
Bump aioaquacell

Jordi1990 committed:
Bump aioaquacell

Jordi1990 committed:
Cleanup and initial tests

Jordi1990 committed:
Fixes for config flow tests

Jordi1990 committed:
Cleanup

Jordi1990 committed:
Formatted

Jordi1990 committed:
Fixes

joostlek commented on homeassistant/components/aquacell/__init__.py:
```diff
@@ -0,0 +1,41 @@
+
+    aquacell_api = AquacellApi(session)
+
+    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = {DATA_AQUACELL: aquacell_api}
+
+    coordinator = AquacellCoordinator(hass, aquacell_api, entry)
+
+    await coordinator.async_config_entry_first_refresh()
+    hass.data[DOMAIN][entry.entry_id] = coordinator
```
You can use [config entry runtime](https://developers.home-assistant.io/blog/2024/04/30/store-runtime-data-inside-config-entry) data instead.

joostlek commented on homeassistant/components/aquacell/config_flow.py:
```diff
@@ -0,0 +1,130 @@
+    api = AquacellApi(session)
+    refresh_token = None
+    try:
+        refresh_token = await api.authenticate(email, password)
```
Is this an oauth2 integration?

Jordi1990 replied:
Unfortunately it's not, it's authenticating through AWS cognito libraries same as the mobile app does. While cognito does also support OAuth2 I haven't managed to get this working.

joostlek commented on homeassistant/components/aquacell/config_flow.py:
```diff
@@ -0,0 +1,130 @@
+                    self.hass, user_input[CONF_EMAIL], user_input[CONF_PASSWORD]
+                )
+                await self.async_set_unique_id(
+                    user_input[CONF_EMAIL].lower(), raise_on_progress=False
```
Can the user change their email address?

Jordi1990 replied:
It looks like that's not possible atm. There's no possibility to do that through the app.

joostlek commented on homeassistant/components/aquacell/config_flow.py:
```diff
@@ -0,0 +1,130 @@
+    }
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
```
```suggestion
class AquaCellConfigFlow(ConfigFlow, domain=DOMAIN):
```

joostlek commented on homeassistant/components/aquacell/config_flow.py:
```diff
@@ -0,0 +1,130 @@
+            errors=errors,
+        )
+
+    async def async_step_reauth(self, user_input=None):
+        """Perform reauth upon an API authentication error."""
+        return await self.async_step_reauth_confirm()
+
+    async def async_step_reauth_confirm(self, user_input=None):
+        """Dialog that informs the user that reauth is required."""
+        errors: dict[str, str] = {}
+
+        if user_input is not None:
+            try:
+                info = await validate_input(
+                    self.hass, user_input[CONF_EMAIL], user_input[CONF_PASSWORD]
+                )
+                entry = await self.async_set_unique_id(
+                    user_input[CONF_EMAIL].lower(), raise_on_progress=False
+                )
+                if entry:
+                    self.hass.config_entries.async_update_entry(entry, data=info)
+                    await self.hass.config_entries.async_reload(entry.entry_id)
+                    return self.async_abort(reason="reauth_successful")
+
+                return self.async_create_entry(title=user_input[CONF_EMAIL], data=info)
+            except CannotConnect:
+                errors["base"] = "cannot_connect"
+            except InvalidAuth:
+                errors["base"] = "invalid_auth"
+            except Exception:  # pylint: disable=broad-except
```
Let's have reauth in the follow up to keep this one small and simple

joostlek commented on homeassistant/components/aquacell/coordinator.py:
```diff
@@ -0,0 +1,85 @@
+
+
+class AquacellCoordinator(DataUpdateCoordinator[list[Softener]]):
+    """My aquacell coordinator."""
+
+    def __init__(
+        self, hass: HomeAssistant, aquacell_api: AquacellApi, entry: ConfigEntry
+    ) -> None:
+        """Initialize coordinator."""
+        super().__init__(
+            hass,
+            _LOGGER,
+            name="Aquacell Coordinator",
+            update_interval=timedelta(seconds=UPDATE_INTERVAL),
+        )
+
+        self.config_entry: ConfigEntry = entry
```
```suggestion
    """My aquacell coordinator."""

    config_entry: ConfigEntry

    def __init__(
        self, hass: HomeAssistant, aquacell_api: AquacellApi
    ) -> None:
        """Initialize coordinator."""
        super().__init__(
            hass,
            _LOGGER,
            name="Aquacell Coordinator",
            update_interval=timedelta(seconds=UPDATE_INTERVAL),
        )
```

joostlek commented on homeassistant/components/aquacell/entity.py:
```diff
@@ -0,0 +1,30 @@
+    """Representation of an aquacell entity."""
+
+    _attr_has_entity_name = True
+    _attr_available = True
```
Why is it always available?

joostlek commented on homeassistant/components/aquacell/entity.py:
```diff
@@ -0,0 +1,30 @@
+    _attr_has_entity_name = True
+    _attr_available = True
+
+    def __init__(self, softener: Softener, coordinator: AquacellCoordinator) -> None:
```
Please put the coordiantor first

joostlek commented on homeassistant/components/aquacell/entity.py:
```diff
@@ -0,0 +1,30 @@
+        """Initialize the aquacell entity."""
+        super().__init__(coordinator)
+
+        self._device_model = softener.dsn
+        self._device_name = softener.name
```
Why are these assigned to `self.` first?

joostlek commented on homeassistant/components/aquacell/manifest.json:
```diff
@@ -0,0 +1,14 @@
+  "name": "Aquacell",
+  "codeowners": ["@Jordi1990"],
+  "config_flow": true,
+  "dependencies": [],
+  "documentation": "https://www.home-assistant.io/integrations/aquacell",
+  "homekit": {},
+  "integration_type": "device",
+  "iot_class": "cloud_polling",
+  "requirements": ["aioaquacell==0.1.4"],
+  "ssdp": [],
+  "zeroconf": []
```
Please remove empty fields.

Also, don't forget to set the `loggers`

joostlek commented on homeassistant/components/aquacell/sensor.py:
```diff
@@ -0,0 +1,157 @@
+class SoftenerEntityDescription(SensorEntityDescription):
+    """Describes Softener sensor entity."""
+
+    value_fn: Callable[[Softener], str | int | float | None]
```
```suggestion
    value_fn: Callable[[Softener], StateType]
```

joostlek commented on homeassistant/components/aquacell/sensor.py:
```diff
@@ -0,0 +1,157 @@
+        key="salt_leftpercentage",
+        translation_key="salt_leftpercentage",
+        state_class=SensorStateClass.MEASUREMENT,
+        icon="mdi:basket-fill",
```
Please use icon translations

joostlek commented on homeassistant/components/aquacell/sensor.py:
```diff
@@ -0,0 +1,157 @@
+        state_class=SensorStateClass.MEASUREMENT,
+        icon="mdi:basket-fill",
+        native_unit_of_measurement=PERCENTAGE,
+        value_fn=lambda softener: cast(int, softener.salt.leftPercent),
```
Does aioaquacell contain enough typing? Consider adding a `py.typed` file to enable mypy to read the `aioaquacell` types

joostlek commented on homeassistant/components/aquacell/sensor.py:
```diff
@@ -0,0 +1,157 @@
+    ),
+    SoftenerEntityDescription(
+        key="salt_rightpercentage",
+        translation_key="salt_rightpercentage",
```
right?

Jordi1990 replied:
This was a bit confusing, the device has salt blocks on the left and right side. I've renamed it to make it more clear.

joostlek commented on homeassistant/components/aquacell/sensor.py:
```diff
@@ -0,0 +1,157 @@
+        native_unit_of_measurement=PERCENTAGE,
+        value_fn=lambda softener: cast(int, softener.salt.rightPercent),
+    ),
+    SoftenerEntityDescription(
+        key="salt_leftdays",
+        translation_key="salt_leftdays",
+        icon="mdi:calendar-today",
+        native_unit_of_measurement=UnitOfTime.DAYS,
+        value_fn=lambda softener: softener.salt.leftDays,
+    ),
+    SoftenerEntityDescription(
+        key="salt_rightdays",
+        translation_key="salt_rightdays",
+        icon="mdi:calendar-today",
+        native_unit_of_measurement=UnitOfTime.DAYS,
+        value_fn=lambda softener: softener.salt.rightDays,
+    ),
```
Could use device class duration

joostlek commented on homeassistant/components/aquacell/sensor.py:
```diff
@@ -0,0 +1,157 @@
+        native_unit_of_measurement=UnitOfTime.DAYS,
+        value_fn=lambda softener: softener.salt.rightDays,
+    ),
+    SoftenerEntityDescription(
+        key="fw_version",
+        translation_key="fw_version",
+        icon="mdi:chip",
+        value_fn=lambda softener: softener.fwVersion,
+    ),
+    SoftenerEntityDescription(
+        key="name",
+        translation_key="name",
+        value_fn=lambda softener: softener.name,
+    ),
```
This is all stored in the device, we should not add this as sensors

joostlek commented on homeassistant/components/aquacell/sensor.py:
```diff
@@ -0,0 +1,157 @@
+        native_unit_of_measurement=PERCENTAGE,
+        value_fn=lambda softener: softener.battery,
+    ),
+    SoftenerEntityDescription(
+        key="last_update",
+        translation_key="last_update",
+        icon="mdi:update",
+        value_fn=lambda softener: datetime.fromtimestamp(
+            cast(float, softener.lastUpdate) / 1000
+        ).isoformat(),
+    ),
```
This is already stored in HA per state, so we should not add this one either

Jordi1990 replied:
This is reported from the device, the time when it has sent an update to the cloud. So when the data is polled you can see if the data is recent or not.

joostlek replied:
Yes, but we don't allow these kinds of sensors in HA as HA already keeps track of that

Jordi1990 replied:
We might have some confusion here, let me explain more thoroughly:
The device sends updates to the cloud, usually once a day. And HA polls the cloud 'cached' values, so it does not retrieve directly from the device. So for example when the salt level stays on 30% for two weeks, you might want to know if it actually measured it recently. HA wouldn't know about this since there's no state change.

joostlek replied:
But you have statistics and can see that the last time the data changed was x days ago.

joostlek replied:
Proposal for now: Omit the sensor, you can add this in a separate PR so you can state your case and it doesn't hold up this PR

Jordi1990 replied:
Ok will do that.

joostlek commented on homeassistant/components/aquacell/sensor.py:
```diff
@@ -0,0 +1,157 @@
+            cast(float, softener.lastUpdate) / 1000
+        ).isoformat(),
+    ),
+    SoftenerEntityDescription(
+        key="wifi_level",
+        translation_key="wifi_level",
+        icon="mdi:wifi",
+        value_fn=lambda softener: softener.wifiLevel,
+    ),
```
What are the possible states?

Jordi1990 replied:
It's reporting in a string type, for example currently mine is 'high'. I haven't mapped out all possible states yet, probably something like high, mid, low. I'll add icon translations for that in a next PR.

joostlek replied:
We should make this an enum sensor which can hold the possible values and allow translatable states.

joostlek commented on homeassistant/components/aquacell/sensor.py:
```diff
@@ -0,0 +1,157 @@
+        icon="mdi:wifi",
+        value_fn=lambda softener: softener.wifiLevel,
+    ),
+    SoftenerEntityDescription(
+        key="lid_in_place",
+        translation_key="lid_in_place",
+        icon="mdi:check",
+        native_unit_of_measurement="",
+        value_fn=lambda softener: softener.lidInPlace,
+    ),
```
Sounds like a true/false sensor, this should be a binary sensor

joostlek commented on homeassistant/components/aquacell/sensor.py:
```diff
@@ -0,0 +1,157 @@
+) -> None:
+    """Set up the sensors."""
+    coordinator: AquacellCoordinator = hass.data[DOMAIN][config_entry.entry_id]
+    entities = [SoftenerSensor(coordinator, sensor) for sensor in SENSORS]
+
+    async_add_entities(entities)
```
```suggestion
    async_add_entities(SoftenerSensor(coordinator, sensor) for sensor in SENSORS)
```

joostlek commented on homeassistant/components/aquacell/sensor.py:
```diff
@@ -0,0 +1,157 @@
+        self.description = description
+        self.softener = self.coordinator.data[0]
+
+        self._attr_translation_key = description.translation_key
+
+        self._attr_unique_id = description.key
+        self._attr_native_unit_of_measurement = description.native_unit_of_measurement
+
+        self._attr_icon = description.icon
+
+        self._attr_state_class = description.state_class
+
+        self._attr_device_class = description.device_class
```
All these can be removed

joostlek commented on homeassistant/components/aquacell/sensor.py:
```diff
@@ -0,0 +1,157 @@
+
+        self._attr_translation_key = description.translation_key
+
+        self._attr_unique_id = description.key
```
We should combine the unique_id with something unique from the device. if I understand someone can add more softeners, so we should add something softener specific to the unique_id

joostlek commented on homeassistant/components/aquacell/sensor.py:
```diff
@@ -0,0 +1,157 @@
+    ) -> None:
+        """Pass coordinator to CoordinatorEntity."""
+        super().__init__(coordinator.data[0], coordinator)
+        self.description = description
```
```suggestion
        self.entity_description = description
```

joostlek commented on homeassistant/components/aquacell/sensor.py:
```diff
@@ -0,0 +1,157 @@
+class SoftenerSensor(AquacellEntity, SensorEntity):
+    """Softener sensor."""
+
+    softener: Softener
```
```suggestion
    entity_description: SoftenerEntityDescription
```

joostlek commented on homeassistant/components/aquacell/sensor.py:
```diff
@@ -0,0 +1,157 @@
+        self._attr_device_class = description.device_class
+
+    @property
+    def native_value(self) -> str | int | float | None:
```
```suggestion
    def native_value(self) -> StateType:
```

joostlek commented on homeassistant/components/aquacell/sensor.py:
```diff
@@ -0,0 +1,157 @@
+        """Return the state of the sensor."""
+        return self.description.value_fn(self.softener)
+
+    @callback
+    def _handle_coordinator_update(self) -> None:
+        """Handle updated data from the coordinator."""
+        self.softener = self.coordinator.data[0]
+        self.async_write_ha_state()
```
```suggestion
    @property
    def softener(self) -> Softener:
        """Handle updated data from the coordinator."""
        return self.coordinator.data[0]
```

joostlek commented on homeassistant/components/aquacell/sensor.py:
```diff
@@ -0,0 +1,157 @@
+    ),
+    SoftenerEntityDescription(
+        key="battery",
+        translation_key="battery",
```
You can omit the battery translations since the device class handles this

joostlek reviewed (changes requested):
Consider adding tests, which would greatly increase the code quality. I have some examples that are very extendible and work really well :)

Jordi1990 committed:
Use config entry runtime
Use icon translations
Removed reauth
Removed last updated sensor
Changed lid in place to binary sensor
Cleanup

Jordi1990 committed:
Remove reauth strings

joostlek commented:
Please keep your first PR to only contain 1 platform, so please move the binary sensor to a follow up

Jordi1990 committed:
Removed binary_sensor platform
Fixed sensors not updating properly

joostlek commented on homeassistant/components/aquacell/entity.py:
```diff
@@ -0,0 +1,35 @@
+    def __init__(
+        self,
+        coordinator: AquacellCoordinator,
+        softener: Softener,
```
Move the softener property to this class instead

joostlek commented on homeassistant/components/aquacell/entity.py:
```diff
@@ -0,0 +1,35 @@
+        self,
+        coordinator: AquacellCoordinator,
+        softener: Softener,
+        entity_description: EntityDescription,
```
I usually just pass a `key` here to do the unique id in the base entity

joostlek commented on homeassistant/components/aquacell/coordinator.py:
```diff
@@ -0,0 +1,84 @@
+        self.password = self.config_entry.data[CONF_PASSWORD]
+        self.aquacell_api = aquacell_api
+
+    async def _async_update_data(self) -> list[Softener]:
```
Okay so there are 2 things that we need to do here:

1. Only keep stuff in the try block that can raise. This function is currently too complex for what it's doing.
2. Instead of returning a list of softeners, transform it into a dict, with the `dsn` as key and the Softener as value. This way you're not bound to the index of a list. I think currently if you would add a third Softener, it could break

joostlek commented on homeassistant/components/aquacell/strings.json:
```diff
@@ -0,0 +1,45 @@
+        "name": "Salt right side days remaining"
+      },
+      "wifi_level": {
+        "name": "WiFi level",
```
```suggestion
        "name": "Wi-Fi level",
```

joostlek commented on homeassistant/components/aquacell/strings.json:
```diff
@@ -0,0 +1,45 @@
+        "name": "Salt right side percentage"
+      },
+      "salt_left_side_remaining_days": {
+        "name": "Salt left side days remaining"
+      },
+      "salt_right_side_remaining_days": {
+        "name": "Salt right side days remaining"
```
I would say time remaining, since we usually don't mention the unit of measurement in the name

joostlek commented on homeassistant/components/aquacell/config_flow.py:
```diff
@@ -0,0 +1,95 @@
+        """Handle the initial step."""
+        errors: dict[str, str] = {}
+        if user_input is not None:
+            try:
```
Only keep stuff in the try block that can raise

joostlek commented on homeassistant/components/aquacell/config_flow.py:
```diff
@@ -0,0 +1,95 @@
+        errors: dict[str, str] = {}
+        if user_input is not None:
+            try:
+                info = await validate_input(
```
I think we can inline this function now (and don't worry I know about the reauth, there are some tricks to reuse this flow to avoid code duplication)

joostlek commented on homeassistant/components/aquacell/sensor.py:
```diff
@@ -0,0 +1,122 @@
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the sensors."""
+    for idx, _ in enumerate(config_entry.runtime_data.data):
+        async_add_entities(
+            SoftenerSensor(config_entry.runtime_data, sensor, idx) for sensor in SENSORS
+        )
```
Please collect the entities in a list before passing them to the async_add_entities. Or use a generator expression

joostlek commented on homeassistant/components/aquacell/sensor.py:
```diff
@@ -0,0 +1,122 @@
+    """Softener sensor."""
+
+    entity_description: SoftenerSensorEntityDescription
+    _softener_index: int
```
Not needed

Jordi1990 replied:
It looks like I do need
`entity_description: SoftenerSensorEntityDescription`
else I get mypy error:
`"SensorEntityDescription" has no attribute "value_fn"`

joostlek replied:
oh I only meant the softener index

joostlek reviewed (changes requested):
Also, if it's ready for review, please use the button for that

Jordi1990 committed:
Remove reauth tests
Bump aioaquacell

Jordi1990 committed:
Moved softener property to entity class
Inlined validate_input method
Renaming of entities
Do a single async_add_entities call to add all entities
Reduced code in try blocks

joostlek commented on homeassistant/components/aquacell/config_flow.py:
```diff
@@ -0,0 +1,78 @@
+                self._abort_if_unique_id_configured()
+                return self.async_create_entry(
+                    title=user_input[CONF_EMAIL],
+                    data={
+                        CONF_REFRESH_TOKEN: refresh_token,
+                        CONF_EMAIL: user_input[CONF_EMAIL],
+                        CONF_PASSWORD: user_input[CONF_PASSWORD],
+                    },
```
```suggestion
                    data={
                        **user_input,
                        CONF_REFRESH_TOKEN: refresh_token,
                    },
```

joostlek commented on homeassistant/components/aquacell/config_flow.py:
```diff
@@ -0,0 +1,78 @@
+            data_schema=DATA_SCHEMA,
+            errors=errors,
+        )
+
+
+class CannotConnect(HomeAssistantError):
+    """Error to indicate we cannot connect."""
+
+
+class InvalidAuth(HomeAssistantError):
+    """Error to indicate there is invalid auth."""
```
Unused now

joostlek commented on homeassistant/components/aquacell/coordinator.py:
```diff
@@ -0,0 +1,87 @@
+        so entities can quickly look up their data.
+        """
+
+        if self.refresh_token is None:
+            raise ConfigEntryError("No token available.")
```
refresh token can never be None

joostlek commented on homeassistant/components/aquacell/coordinator.py:
```diff
@@ -0,0 +1,87 @@
+            _LOGGER.debug("Logging in using: %s", self.refresh_token)
+
+            try:
+                await self.aquacell_api.authenticate_refresh(self.refresh_token)
```
What's the validity of a refresh token? Can we store that as well somewhere?

Jordi1990 replied:
It's currently set to 30 days, depends on their AWS configuration. Haven't seen this value in any of the authentication responses so decided to just attempt to reauthenticate whenever it fails to authenticate using the refresh token.

We could hardcode the 30 days and check on that, but it wouldn't be robuust against configuration changes on their cloud.

joostlek replied:
What kind of token is it? a JWT?

Jordi1990 replied:
Yes, but the payload is encrypted.

joostlek replied:
Do you have an example?

Jordi1990 replied:
`eyJjdHkiOiJKV1QiLCJlbmMiOiJBMjU2R0NNIiwiYWxnIjoiUlNBLU9BRVAifQ.QZu553nflsb4EGa0mje9V8ySxd62MSwF6ceNpZnCvi_oTs3kIOdyoA7tbSbXaiNmlm1Rrfj-lMqJ36Cc7KKRrIfAKVvOnw6uccBYjwS-E9pn0hWQhlanLA8zPwlxNi9_lt2yhcNCvFvPEwFzmzj_-n7bOXvq0LNcNurv7tpdfA6ftfqvGj2vQHMxzIsDZ0KzrOsZlxEXoVfxuvk12RhwM1AG-oBJZ3FXpjGAnim-YIVNKWWRX35tUV4-Qhl5oMXPjb5ccSss-ONE7Ohkd-leC9eDhIw7osU_Ny6aP1kSOOO1qwq0exXOJwWWFqL_cHmL22Y80AcXdpHZAjJW7go_UA`

joostlek replied:
Oh lol, i was like oh it's just a jwt, looks usable. But then indeed. Why would you encrypt this haha.

Hmm, i think we can maybe just roll with storing the date we requested the token and check if it exceeds 30 days. This makes the check if the token is valid a lot cheaper. It would only become a real problem when they make the validity 1 day for example. It would trigger reauth every day, but given that people watch issues and we release a hotfix about every week, we can then change this.

Wdyt?

joostlek replied:
Oh wait, I forgot that we already have username and password, so we don't need to do reauth, unless the token retrieval changes. So I think this can be a very user friendly approach

Jordi1990 replied:
Yeah agreed, I'll add the 30 day check.

joostlek commented on homeassistant/components/aquacell/coordinator.py:
```diff
@@ -0,0 +1,87 @@
+        try:
+            softeners = await self.aquacell_api.get_all_softeners()
+        except AuthenticationFailed as err:
+            raise ConfigEntryAuthFailed from err
```
```suggestion
            raise ConfigEntryError from err
```
Because we don't have reauth yet

joostlek commented on homeassistant/components/aquacell/entity.py:
```diff
@@ -0,0 +1,41 @@
+
+        self.softener_key = softener_key
+
+        self._attr_unique_id = f"{self.softener.dsn}-{entity_key}"
```
```suggestion
        self._attr_unique_id = f"{softener_key}-{entity_key}"
```

joostlek commented on homeassistant/components/aquacell/entity.py:
```diff
@@ -0,0 +1,41 @@
+        self._attr_device_info = DeviceInfo(
+            name=self.softener.name,
+            hw_version=self.softener.fwVersion,
+            identifiers={(DOMAIN, str(self.softener.dsn))},
+            manufacturer=self.softener.brand,
+            model=self.softener.ssn,
+            serial_number=self.softener.dsn,
```
```suggestion
            identifiers={(DOMAIN, str(softener_key))},
            manufacturer=self.softener.brand,
            model=self.softener.ssn,
            serial_number=softener_key,
```

joostlek commented on homeassistant/components/aquacell/__init__.py:
```diff
@@ -0,0 +1,40 @@
+
+    aquacell_api = AquacellApi(session)
+
+    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = {DATA_AQUACELL: aquacell_api}
```
We don't need this now

joostlek commented on tests/components/aquacell/test_config_flow.py:
```diff
@@ -0,0 +1,111 @@
+pytestmark = pytest.mark.usefixtures("mock_setup_entry")
+
+
+async def __mock_authenticate(self, user_name, password) -> string:
+    return "refresh-token"
```
Please create a test fixture for the `AquacellApi`, check `epion`, `analytics_insights` for examples

joostlek commented on tests/components/aquacell/test_config_flow.py:
```diff
@@ -0,0 +1,111 @@
+async def test_form(hass: HomeAssistant, mock_setup_entry: AsyncMock) -> None:
+    """Test we get the form."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
```
```suggestion
        DOMAIN, context={"source": SOURCE_USER}
```

joostlek commented on tests/components/aquacell/test_config_flow.py:
```diff
@@ -0,0 +1,111 @@
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+    assert result["type"] == FlowResultType.FORM
```
```suggestion
    assert result["type"] is FlowResultType.FORM
```
Please change this in all your tests

joostlek commented on tests/components/aquacell/test_config_flow.py:
```diff
@@ -0,0 +1,111 @@
+    assert len(mock_setup_entry.mock_calls) == 1
+
+
+async def test_form_invalid_auth(hass: HomeAssistant) -> None:
+    """Test we handle invalid auth."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+
+    with patch(
+        "homeassistant.components.aquacell.config_flow.AquacellApi.authenticate",
+        side_effect=AuthenticationFailed,
+    ):
+        result2 = await hass.config_entries.flow.async_configure(
+            result["flow_id"],
+            {
+                CONF_EMAIL: "test@test.com",
+                CONF_PASSWORD: "test-password",
+            },
+        )
+
+    assert result2["type"] == FlowResultType.FORM
+    assert result2["errors"] == {"base": "invalid_auth"}
+
+
+async def test_form_cannot_connect(hass: HomeAssistant) -> None:
+    """Test we handle cannot connect error."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
```
1. You can use parametrize to combine these tests and make the exception and the error string a test parameter
2. Both tests should end in `CREATE_ENTRY` to make sure the entry can succeed after hitting an error

joostlek commented on tests/components/aquacell/test_config_flow.py:
```diff
@@ -0,0 +1,111 @@
+    assert result2["errors"] == {"base": "cannot_connect"}
+
+
+async def test_form_unexpected_exception(hass: HomeAssistant) -> None:
+    """Test we handle unexpected exception."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
```
This one can be parametrized as well

Jordi1990 committed:
Made tests parameterized and use test fixture for api
Cleaned up unused code
Removed traces of reauth

Jordi1990 committed:
Add check if refresh token is expired
Add tests

Jordi1990 commented:
@joostlek Ready to review, wasn't sure if you've noticed.

joostlek commented:
Oh i did not

joostlek commented on homeassistant/components/aquacell/coordinator.py:
```diff
@@ -0,0 +1,94 @@
+
+        return {softener.dsn: softener for softener in softeners}
+
+    async def _reauthenticate(self) -> None:
+        _LOGGER.debug("Attempting to renew refresh token")
+        refresh_token = await self.aquacell_api.authenticate(self.email, self.password)
+        self.refresh_token = refresh_token
+        self._update_config_entry_refresh_token()
+
+    def _update_config_entry_refresh_token(self) -> None:
+        """Update or delete the refresh_token in the Config Entry."""
+        data = {
+            **self.config_entry.data,
+            CONF_REFRESH_TOKEN: self.refresh_token,
+            CONF_REFRESH_TOKEN_CREATION_TIME: datetime.now().timestamp(),
+        }
+
+        self.hass.config_entries.async_update_entry(self.config_entry, data=data)
```
This can be merged

joostlek commented on tests/components/aquacell/test_config_flow.py:
```diff
@@ -0,0 +1,84 @@
+
+from tests.components.aquacell import TEST_CONFIG_ENTRY, TEST_USER_INPUT
+
+pytestmark = pytest.mark.usefixtures("mock_setup_entry")
```
Can be removed

joostlek commented on tests/components/aquacell/test_sensor.py:
```diff
@@ -0,0 +1,36 @@
+from tests.components.aquacell import DSN, setup_integration
+
+
+@pytest.mark.parametrize(
+    ("sensor_id"),
+    [
+        ("wi_fi_level"),
+    ],
+)
+async def test_sensors(
+    hass: HomeAssistant,
+    mock_aquacell_api: AsyncMock,
+    mock_config_entry: MockConfigEntry,
+    entity_registry: er.EntityRegistry,
+    sensor_id: str,
+) -> None:
+    """Test the creation of Aquacell sensors."""
+    await setup_integration(hass, mock_config_entry)
+    entity_id = f"sensor.aquacell_name_{sensor_id}"
+
+    assert entity_registry.async_is_registered(entity_id)
+    entry = entity_registry.async_get(entity_id)
+    assert entry.unique_id == f"{DSN}-{sensor_id}"
+    assert hass.states.get(entity_id).state == "high"
```
While you're at it, might as well use snapshots to fixate all entities created for the sensor platform, checkout `flexit_bacnet` or `analytics_insights`

Jordi1990 replied:
Done

Jordi1990 committed:
Add missing unique_id to config entry mock
Inlined _update_config_entry_refresh_token method
Fix incorrect test method name and comment

Jordi1990 committed:
Add snapshot test
Changed WiFi level to WiFi strength

joostlek commented:
Putting to draft as there are still comments open

Jordi1990 committed:
Bump aioaquacell to 0.1.7

joostlek commented on homeassistant/components/aquacell/config_flow.py:
```diff
@@ -0,0 +1,70 @@
+                _LOGGER.exception("Unexpected exception")
+                errors["base"] = "unknown"
+            else:
+                await self.async_set_unique_id(
+                    user_input[CONF_EMAIL].lower(), raise_on_progress=False
+                )
+                self._abort_if_unique_id_configured()
```
We can put this directly after `if user_input is not None:`

joostlek commented on tests/components/aquacell/test_config_flow.py:
```diff
@@ -0,0 +1,82 @@
+"""Test the Aquacell config flow."""
```
I'm missing a test for the duplicate entry abort

joostlek commented on tests/components/aquacell/test_coordinator.py:
```diff
@@ -0,0 +1,85 @@
+"""Test the Aquacell coordinator."""
```
I personally dislike `test_coordinator.py` tests as they assume you use a coordinator. But they all look like general integration runtime things so they could look good in `test_init.py`

Jordi1990 replied:
Moved them to test_init.py

joostlek commented on tests/components/aquacell/test_init.py:
```diff
@@ -0,0 +1,29 @@
+) -> None:
+    """Test load and unload entry."""
+    await setup_integration(hass, mock_config_entry)
+    entry = hass.config_entries.async_entries(DOMAIN)[0]
```
You can just use `mock_config_entry` here, it's the same object as you're fetching now

Jordi1990 committed:
Move test_coordinator tests to test_init
Add test for duplicate config entry
